File: javascript/MergeSort.js
Merge = function(array) {
    if (array.length <= 1) {
        return array
    }

    half = parseInt(array.length / 2);
    lefthalf = array.slice(0, half);
    righthalf = array.slice(half, array.length);
    return MergeSort(Merge(lefthalf), Merge(righthalf));
}

MergeSort = function(left_arr, right_arr) {
    result = [];
    while ((left_arr.length > 0) && (right_arr.length > 0)) {
        if (left_arr[0] < right_arr[0]) {
            result.push(left_arr.shift());
        } else {
            result.push(right_arr.shift());
        }
    }

    while (left_arr.length > 0) {
        result.push(left_arr.shift());
    }

    while (right_arr.length > 0) {
        result.push(right_arr.shift());
    }

    console.log("SORT RESULT: " + result); // Debug
    return result;
}

// Function for checking equality of arrays that
// don't contain nested arrays
Array.prototype.equals = function(array) {
    if (!array) { return false; }
    if (array.length !== this.length) { return false; }

    // Check for equality
    for (x = 0; x < array.length; x++) {
        if (this[x] !== array[x]) { return false; }
    }
    return true;
}

// Tests
var results = {successes: 0, failures: 0};
testMergeSort = function(input) {
    output = Merge(input.concat());
    if (output.equals(input.sort())) {
        results.successes++;
        console.log(`Success: ${output}`);
    } else {
        results.failures++;
        console.log(`Failure: ${output}`);
    }
}

inputs = [[], [2, 1], [1, 2, 3, 4, 5], [5, 4, 3, 2, 1], [100, 1000, 10], [-1, -10, 4, 0], [1]];
for (x = 0; x < inputs.length; x++) {
    testMergeSort(inputs[x]);
}
console.log(`${results.successes} successes, ${results.failures} failures.`);

File: python/MergeSort.py
def mergeSort(alist):
    if len(alist)>1:
        mid = len(alist)//2
        lefthalf = alist[:mid]
        righthalf = alist[mid:]

        mergeSort(lefthalf)
        mergeSort(righthalf)

        i=0
        j=0
        k=0
        while i < len(lefthalf) and j < len(righthalf):
            if lefthalf[i] < righthalf[j]:
                alist[k]=lefthalf[i]
                i+=1
            else:
                alist[k]=righthalf[j]
                j+=1
            k+=1

        while i < len(lefthalf):
            alist[k]=lefthalf[i]
            i+=1
            k+=1

        while j < len(righthalf):
            alist[k]=righthalf[j]
            j+=1
            k+=1

if __name__ == "__main__":
    import unittest
    class MergeSortTest(unittest.TestCase):
        def test_merge_sort(self):
            inputs = ([54,26,93,17,77,31,44,55,20], [])
            for result in inputs:
                mergeSort(i)
                expected = sorted(i)
                self.assertTrue(result == expected)

        unittest.main()
